R/calc_coherence.R

```diff
@@ -0,0 +1,259 @@
+    #NORMALIZED PMI
+    #again, in contrast, to other implementations, only intrinsic NPMI as in PMIM (for implementation with sliding window see, e.g., Bouma, 2009)
+    ,npmi = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant) {(log2((tcm[wi,wj]/ndocs_tcm) + log_smooth_constant) - log2(tcm[wi,wi]/ndocs_tcm) - log2(tcm[wj,wj]/ndocs_tcm)) /  -log2((tcm[wi,wj]/ndocs_tcm) + log_smooth_constant)}
+                      ,set_type = "one_pre"
+                      ,aggr_fun = "function(x) mean(x, na.rm = T)")
+    #DIFFERENCE
+    #assuming we use ordered tcm it follows that p(wi)>p(wj)
+    #to set bounds of the measures [-1,1] (1 is good)  wi/wj are switched in formula
+    #this is similar (not exactly the same) to the measure of textmineR package https://github.com/TommyJones/textmineR/issues/35
+    ,prob_dif = structure(function(wi, wj, ndocs_tcm, tcm, log_smooth_constant) {tcm[wj,wi]/tcm[wj,wj] - (tcm[wj,wj]/ndocs_tcm)}
+                          ,set_type = "one_suc"
+                          ,aggr_fun = "function(x) mean(x, na.rm = T)")
+  )
+
+#WRAPPER FUNCTION TO CALCULATE COHERENCE-------------------------------------------------------
+  #TODO coh_funs and other args need to be passed to the function as argument,
+  #later steps would be less verbose when fetching several or the arguments,e.g., tcm, from the function environment -> scoping
+  calc_coh <- function( idxs
+                       , tcm
+                       , coh_funs = coh_funs
+                       , ndocs_tcm
+                       , coh_measure
+                       , log_smooth_constant = log_smooth_constant
+                       , alternative_order
+                         ) {
+    #select coherence function from the ones availble
+    coh_fun <- coh_funs[[coh_measure]]
+    #define the wi wj set
+    coh <- as.data.table(create_wiwj_set(idxs, set_type = attr(coh_fun, "set_type"), alternative_order = alternative_order))
+    #calculate score for each pair of wi wj
+    coh[, coh_res:= mapply(function(x,y) coh_fun(x,y, tcm = tcm, ndocs_tcm = ndocs_tcm, log_smooth_constant = log_smooth_constant),wi, wj)]
+    #aggregate
+    aggr_fun <- eval(parse(text = attr(coh_fun, "aggr_fun")))
+    coh[, round(aggr_fun(coh_res), d = 4)]
+  }
+
+#CALCULATE COHERENCE----------------------------------------------------------------
+  for (i in names(coh_funs)) {
+    topic_coherence[,(i):= lapply(tcm_idxs, function(x) {
+                                     calc_coh( idxs = x, coh_measure = i
+                                    , tcm = tcm
+                                    , coh_funs = coh_funs
+                                    , ndocs_tcm = ndocs_tcm
+                                    , log_smooth_constant = log_smooth_constant
+                                    , alternative_order = restore_topic_order)})
+                    , by = Topic]
+  }
+
+  topic_coherence[, tcm_idxs := NULL]
+
+  if (average_over_topics == TRUE) {
```

Comment:
not sure this is needed - user can calculate this easily himself

Reply:
...done via a4d6a22, also see above.